database/src/database.rs

```diff
@@ -39,12 +46,73 @@ pub trait DatabaseRo<T: Table> {
     /// if a particular key in the `range` does not exist,
     /// [`RuntimeError::KeyNotFound`] wrapped in [`Err`] will be returned
     /// from the iterator.
+    #[allow(clippy::iter_not_returning_iterator)]
     fn get_range<'a, Range>(
         &'a self,
         range: Range,
-    ) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + 'a, RuntimeError>
+    ) -> Result<impl Iterator<Item = Result<(T::Key, T::Value), RuntimeError>> + 'a, RuntimeError>
     where
         Range: RangeBounds<T::Key> + 'a;
+
+    /// TODO
+    ///
+    /// # Errors
+    /// TODO
+    #[allow(clippy::iter_not_returning_iterator)]
+    fn iter(
+        &self,
+    ) -> Result<impl Iterator<Item = Result<(T::Key, T::Value), RuntimeError>> + '_, RuntimeError>;
+
+    /// TODO
+    ///
+    /// # Errors
+    /// TODO
+    fn keys(&self)
+        -> Result<impl Iterator<Item = Result<T::Key, RuntimeError>> + '_, RuntimeError>;
+
+    /// TODO
+    ///
+    /// # Errors
+    /// TODO
+    fn values(
+        &self,
+    ) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + '_, RuntimeError>;
+
+    /// TODO
+    ///
+    /// # Errors
+    /// TODO
+    fn contains(&self, key: &T::Key) -> Result<bool, RuntimeError> {
+        match self.get(key) {
+            Ok(_) => Ok(true),
+            Err(RuntimeError::KeyNotFound) => Ok(false),
+            Err(e) => Err(e),
+        }
+    }
```

Comment:
I think `RuntimeError::KeyNotFound` shouldn't exist and we should return `Option<(T::Key, T::Value)>` like all other APIs do. On stuff like `get()` too.

Reply:
With the service interface this be an error any way, so we would have to map the option to an error. IMO I don't think this is worth it

Reply:
The main reason is because `?` no longer works for callers, aka, any non-`service` user.

You can no longer use `?` because the error might be a non-existent key (not really a hard error) or an _actual_ database error, so all callers must:
```rust
fn get_block(db: &Database) -> Result<Option<Block>, Error> {
    match db.get(&key) {
        Ok(v) => Ok(Some(v)),
        Err(RuntimeError::KeyNotFound) => Ok(None),
        Err(e) => Err(e),
    }
}
```
`Option` is probably used in most APIs because it represents the problem more accurately and allows `?`:
```rust
fn get_block(db: &Database) -> Result<Option<Block>, Error> {
    Ok(db.get(&key)?)
}
```
With `Option`, yes `cuprate_database::service` will have to map it but taking away `?` from non-`service` users and forcing `match` at all call sites is also quite bad. Although, I guess it does make usage within Cuprate easier.

Reply:
I would rather keep it an error.

We will still need the error variant for `KeyNotFound` for the service and we will be wrapping the return in an option to then "unwrap" it. 

Up to you though.

Reply:
Will keep it as is since both have tradeoffs - no `Option`.